File: samples/java/spark/sources/SparkCovarianceDense.java
/* file: SparkCovarianceDense.java */
/*******************************************************************************
* Copyright 2017-2019 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
//  Content:
//      Java sample of dense variance-covariance matrix computation in the
//      distributed processing mode
////////////////////////////////////////////////////////////////////////////////
*/

package DAAL;

import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;
import java.util.Map;
import java.util.HashMap;

import org.apache.spark.api.java.*;
import org.apache.spark.api.java.function.*;
import org.apache.spark.SparkConf;

import com.intel.daal.algorithms.covariance.*;
import com.intel.daal.data_management.data.*;
import com.intel.daal.services.*;

public class SparkCovarianceDense {
    /* Class containing the algorithm results */
    static class CovarianceResult {
        public HomogenNumericTable covariance;
        public HomogenNumericTable mean;
    }

    public static CovarianceResult runCovariance(DaalContext context, JavaRDD<HomogenNumericTable> dataRDD) {
        JavaRDD<PartialResult> partsRDD = computeStep1Local(context, dataRDD);

        PartialResult reducedPres = reducePartialResults(context, partsRDD);

        CovarianceResult result = finalizeMergeOnMasterNode(context, reducedPres);

        return result;
    }

    private static JavaRDD<PartialResult> computeStep1Local(DaalContext context, JavaRDD<HomogenNumericTable> dataRDD) {
        return dataRDD.map(new Function<HomogenNumericTable, PartialResult>() {
            public PartialResult call(HomogenNumericTable table) {
                DaalContext localContext = new DaalContext();

                /* Create an algorithm to compute a dense variance-covariance matrix on local nodes*/
                DistributedStep1Local covarianceLocal = new DistributedStep1Local(localContext, Double.class, Method.defaultDense);

                /* Set the input data on local nodes */
                table.unpack(localContext);
                covarianceLocal.input.set(InputId.data, table);

                /* Compute a dense variance-covariance matrix on local nodes */
                PartialResult pres = covarianceLocal.compute();
                pres.pack();

                localContext.dispose();
                return pres;
            }
        });
    }

    private static PartialResult reducePartialResults(DaalContext context, JavaRDD<PartialResult> partsRDD) {
        return partsRDD.reduce(new Function2<PartialResult, PartialResult, PartialResult>() {
            public PartialResult call(PartialResult pr1, PartialResult pr2) {
                DaalContext localContext = new DaalContext();

                /* Create an algorithm to compute new partial result from two partial results */
                DistributedStep2Master covarianceMaster = new DistributedStep2Master(localContext, Double.class, Method.defaultDense);

                /* Set the input data recieved from the local nodes */
                pr1.unpack(localContext);
                pr2.unpack(localContext);
                covarianceMaster.input.add(DistributedStep2MasterInputId.partialResults, pr1);
                covarianceMaster.input.add(DistributedStep2MasterInputId.partialResults, pr2);

                /* Compute a new partial result from two partial results */
                PartialResult reducedPresLocal = (PartialResult)covarianceMaster.compute();
                reducedPresLocal.pack();

                localContext.dispose();
                return reducedPresLocal;
            }
        });
    }

    private static CovarianceResult finalizeMergeOnMasterNode(DaalContext context, PartialResult reducedPres) {

        /* Create an algorithm to compute a dense variance-covariance matrix on the master node */
        DistributedStep2Master covarianceMaster = new DistributedStep2Master(context, Double.class, Method.defaultDense);

        /* Set the reduced partial result to the master algorithm to compute the final result */
        reducedPres.unpack(context);
        covarianceMaster.input.add(DistributedStep2MasterInputId.partialResults, reducedPres);

        /* Compute a dense variance-covariance matrix on the master node */
        covarianceMaster.compute();

        /* Finalize computations and retrieve the results */
        Result res = covarianceMaster.finalizeCompute();

        CovarianceResult covResult = new CovarianceResult();
        covResult.covariance = (HomogenNumericTable)res.get(ResultId.covariance);
        covResult.mean = (HomogenNumericTable)res.get(ResultId.mean);
        return covResult;
    }
}

File: samples/python/spark/sources/spark_CovarianceDense.py
# file: spark_CovarianceDense.py
#===============================================================================
# Copyright 2017-2019 Intel Corporation.
#
# This software and the related documents are Intel copyrighted  materials,  and
# your use of  them is  governed by the  express license  under which  they were
# provided to you (License).  Unless the License provides otherwise, you may not
# use, modify, copy, publish, distribute,  disclose or transmit this software or
# the related documents without Intel's prior written permission.
#
# This software and the related documents  are provided as  is,  with no express
# or implied  warranties,  other  than those  that are  expressly stated  in the
# License.
#
# License:
# http://software.intel.com/en-us/articles/intel-sample-source-code-license-agr
# eement/
#===============================================================================

#
#  Content:
#     Python sample of dense variance-covariance matrix computation
#
from __future__ import print_function
import os
import sys

from pyspark import SparkContext, SparkConf

from daal import step1Local, step2Master
from daal.algorithms import covariance
from daal.data_management import OutputDataArchive

from distributed_hdfs_dataset import DistributedHDFSDataSet
from distributed_hdfs_dataset import serializeNumericTable
from distributed_hdfs_dataset import deserializeNumericTable

utils_folder = os.path.join(os.environ['DAALROOT'], 'examples', 'python', 'source')
if utils_folder not in sys.path:
    sys.path.insert(0, utils_folder)
from utils import printNumericTable


def runCovariance(dataRDD):
    partsRDD = computestep1Local(dataRDD)
    return finalizeMergeOnMasterNode(partsRDD)


def computestep1Local(dataRDD):

    def mapper(tup):

        key, val = tup

        # Create an algorithm to compute a dense variance-covariance matrix on local nodes
        covarianceLocal = covariance.Distributed(step=step1Local, method=covariance.defaultDense)

        # Set the input data on local nodes
        deserialized_val = deserializeNumericTable(val)
        covarianceLocal.input.set(covariance.data, deserialized_val)

        # Compute a dense variance-covariance matrix on local nodes
        pres = covarianceLocal.compute()

        serialized_pres = serializeNumericTable(pres)

        return (key, serialized_pres)

    return dataRDD.map(mapper)


def finalizeMergeOnMasterNode(partsRDD):

    # Create an algorithm to compute a dense variance-covariance matrix on the master node
    covarianceMaster = covariance.Distributed(step=step2Master, method=covariance.defaultDense)

    parts_list = partsRDD.collect()

    # Add partial results computed on local nodes to the algorithm on the master node
    for _, val in parts_list:
        dataArch = OutputDataArchive(val)
        deserialized_val = covariance.PartialResult()
        deserialized_val.deserialize(dataArch)
        covarianceMaster.input.add(covariance.partialResults, deserialized_val)

    # Compute a dense variance-covariance matrix on the master node
    covarianceMaster.compute()

    # Finalize computations and retrieve the results
    res = covarianceMaster.finalizeCompute()

    result = {}
    result['covariance'] = res.get(covariance.covariance)
    result['mean'] = res.get(covariance.mean)

    return result

if __name__ == "__main__":

    # Create SparkContext that loads defaults from the system properties and the classpath and sets the name
    sc = SparkContext(conf=SparkConf().setAppName("Spark covariance(dense)").setMaster("local[4]"))

    # Read from the distributed HDFS data set at a specified path
    dd = DistributedHDFSDataSet("/Spark/CovarianceDense/data/")
    dataRDD = dd.getAsPairRDD(sc)

    # Redirect stdout to a file for correctness verification
    stdout = sys.stdout
    sys.stdout = open('CovarianceDense.out', 'w')

    # Compute a dense variance-covariance matrix for dataRDD
    final_result = runCovariance(dataRDD)

    # Print the results
    printNumericTable(final_result['covariance'], "Covariance:", interval=9)
    printNumericTable(final_result['mean'], "Mean:", interval=9)

    # Restore stdout
    sys.stdout = stdout

    sc.stop()
